### ISSUE TITLE
[FEATURE REQUEST] return result to middleware

### ISSUE THREAD

**Issue by ChristianYeah** [hibiken/asynq#218]
in some circumstances, a chained job task is required.

I prefer not to write the logic in the current task to re-queue the post-process task

however, the payload is not mutable and only errors will return by the task handler, the middleware is not able to catch any result directly(except the Redis, etc.)

I prefer to consider this as a discussion rather than a feature request, I'm not pretty professional in the system's structure design.

thank you

**Comment by hibiken** [comment]
Hi @ChristianYeah! Thank you for starting this discussion.

I believe this is a common need, to create some kind of a workflow of sort. 
Examples:
- TaskB should be processed only after TaskA has been successfully processed.
- TaskD should be processed only after TaskA, TaskB, TaskC have been successfully processed.

---

Do you have ideas on how the API should look like from the user of the library point of view?

**Comment by sujit-baniya** [comment]
@hibiken There's a library for workflow: https://github.com/rdleal/go-async

Maybe the `Auto` part is something you might consider.
```go
package main

import (
        "context"
        "encoding/json"
        "fmt"
        "log"
        "strings"

        "github.com/rdleal/go-async/async"
)

func main() { 
	type msg struct {
		Greet, Name string
	}

	const jsonStr = `{"greet": "hello", "name" : "gopher"}`
	fnc, err := async.Auto(context.Background(), async.FuncMap{
		"reader":  strings.NewReader,
		"decoder": async.DependsOn("reader").To(json.NewDecoder),
		"parser": async.DependsOn("decoder").To(func(d *json.Decoder) (msg, error) {
			var m msg
			err := d.Decode(&m)
			return m, err
		}),
		"greet": async.DependsOn("parser").To(func(m msg) string {
			return strings.Title(m.Greet)
		}),
		"name": async.DependsOn("parser").To(func(m msg) string {
			return strings.Title(m.Name)
		}),
		"greeter": async.DependsOn("greet", "name").To(func(greet, name string) string {
			return fmt.Sprintf("%s, %s!", greet, name)
		}),
	}, async.WithArgs(jsonStr))
	if err != nil {
		log.Fatal(err)
	}

	for fn := range fnc {
		res, err := fn.Returned()
		if err != nil {
			log.Fatal(err)
		}
		fmt.Println(res[0]) // Hello, Gopher!
	}
} 
```

**Comment by sujit-baniya** [comment]
Also in my view, for this library we need 
- to have some kind of event handling based on task state.
- pass result of one task to another task

**Comment by ChristianYeah** [comment]
> Hi @ChristianYeah! Thank you for starting this discussion.
> 
> I believe this is a common need, to create some kind of a workflow of sort.
> Examples:
> 
> * TaskB should be processed only after TaskA has been successfully processed.
> * TaskD should be processed only after TaskA, TaskB, TaskC have been successfully processed.
> 
> Do you have ideas on how the API should look like from the user of the library point of view?

I think the simplest way is to allow the worker&task handler redirects its result to middleware or post-process middleware.
So developers could define the workflow in the payload rather than in golang code(rebuild & re-deploy are needed). users just need to ensure it's a non-cyclic graph or a DAG.

**Comment by hibiken** [comment]
@ChristianYeah Do you think storing a result in the context would work? Using [`WithValue`](https://golang.org/pkg/context/#WithValue) from context package. 

I'm interested in building a feature for this in a longer term. I'll study how [Celery](https://docs.celeryproject.org/en/stable/userguide/canvas.html) and [Sidekiq](https://github.com/mperham/sidekiq/wiki/Batches) is handling this type of workflow.

**Comment by ChristianYeah** [comment]
@hibiken  I tried to override the context via ```WithValue``` function in task's ```ProcessTask``` function. however, I'm not able to get the value in the context in the middleware.

**Comment by hibiken** [comment]
Yes, you are right, it's not possible to store a result in context and make it available to middleware since context is immutable. 
This may require API change in `Handler` interface to return a result as well as an error. I'll work in the next iteration after I wrap up my current work on web ui. 

In the meantime, maybe you could store the result in redis and access it in middleware?

**Comment by hibiken** [comment]
I'll move the discussion to #244 and close this one.
